api/src/test/java/io/strimzi/api/kafka/model/ApiEvolutionCrdIT.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    private Long createOrReplaceCrd(String s) throws IOException {
+        File src = new File(s);
+        if (!src.exists()) {
+            throw new RuntimeException(src.getAbsolutePath() + " does not exist");
+        }
+        CustomResourceDefinition crd = new YAMLMapper().readValue(src, CustomResourceDefinition.class);
+        LOGGER.info("Create or replacing {} with versions {}", crd.getMetadata().getName(),
+                crd.getSpec().getVersions().stream()
+                        .map(v -> v.getName() + "{stored=" + v.getStorage() + "}")
+                        .collect(Collectors.joining(", ")));
+        return cluster.client().getClient().customResourceDefinitions().createOrReplace(crd).getMetadata().getGeneration();
+    }
+
+    private void deleteCrd(String s) throws IOException {
+        File src = new File(s);
+        if (!src.exists()) {
+            throw new RuntimeException(src.getAbsolutePath() + " does not exist");
+        }
+        CustomResourceDefinition crd = new YAMLMapper().readValue(src, CustomResourceDefinition.class);
+        LOGGER.info("Create or replacing {} with versions {}", crd.getMetadata().getName(),
+                crd.getSpec().getVersions().stream()
+                        .map(v -> v.getName() + "{stored=" + v.getStorage() + "}")
+                        .collect(Collectors.joining(", ")));
+        cluster.client().getClient().customResourceDefinitions().delete(crd);
+    }
+
+    private void assertIsMapListener(Kafka kafka) {
+        Assertions.assertNotNull(kafka);
+        Assertions.assertNotNull(kafka.getSpec());
+        Assertions.assertNotNull(kafka.getSpec().getKafka());
+        Assertions.assertNotNull(kafka.getSpec().getKafka().getListeners());
+        Assertions.assertNotNull(kafka.getSpec().getKafka().getListeners().getKafkaListeners());
+        Assertions.assertNull(kafka.getSpec().getKafka().getListeners().getGenericKafkaListeners());
+    }
+
+    private void assertIsListListener(Kafka kafka) {
+        Assertions.assertNotNull(kafka);
+        Assertions.assertNotNull(kafka.getSpec());
+        Assertions.assertNotNull(kafka.getSpec().getKafka());
+        Assertions.assertNotNull(kafka.getSpec().getKafka().getListeners());
+        Assertions.assertNull(kafka.getSpec().getKafka().getListeners().getKafkaListeners());
+        Assertions.assertNotNull(kafka.getSpec().getKafka().getListeners().getGenericKafkaListeners());
+    }
+
+    private NonNamespaceOperation<Kafka, KafkaList, DoneableKafka, Resource<Kafka, DoneableKafka>> v1beta1Op() {
+        return Crds.kafkaV1Beta1Operation(cluster.client().getClient()).inNamespace(NAMESPACE);
+    }
+
+    private NonNamespaceOperation<Kafka, KafkaList, DoneableKafka, Resource<Kafka, DoneableKafka>> v1beta2Op() {
+        return Crds.kafkaV1Beta2Operation(cluster.client().getClient()).inNamespace(NAMESPACE);
+    }
+
+    private GenericKafkaListener listListener() {
```

Comment:
Wouldn't be better to name it createInternalPlainListener or am I missing something?

Reply:
The precise details of the listener don't matter (the CR is never actually consumed by the operator), instead what we're testing here is that we API server lets us create a Kafka CR with listeners as a list rather than a map.